include/merlin/core_kernels/find_or_insert.cuh

```diff
@@ -21,6 +21,1253 @@
+      } while (!result);
+      if (result) {
+        occupy_result = OccupyResult::DUPLICATE;
+        key_pos = possible_pos;
+        S* src = BUCKET::scores(bucket_keys_ptr, bucket_capacity, key_pos);
+        score = *src;
+        break;
+      } else if (bucket_size == bucket_capacity) {
+        continue;
+      }
+      VecD_Comp empty_digests_ = empty_digests<K>();
+      cmp_result = __vcmpeq4(probe_digests, empty_digests_);
+      cmp_result &= 0x01010101;
+      do {
+        if (cmp_result == 0) break;
+        uint32_t index = (__ffs(cmp_result) - 1) >> 3;
+        cmp_result &= (cmp_result - 1);
+        possible_pos = pos_cur + i * 4 + index;
+        if (offset == 0 && possible_pos < key_pos) continue;
+        auto current_key = BUCKET::keys(bucket_keys_ptr, possible_pos);
+        K expected_key = static_cast<K>(EMPTY_KEY);
+        result = current_key->compare_exchange_strong(
+            expected_key, static_cast<K>(LOCKED_KEY),
+            cuda::std::memory_order_acquire, cuda::std::memory_order_relaxed);
+      } while (!result);
+      if (result) {
+        occupy_result = OccupyResult::OCCUPIED_EMPTY;
+        key_pos = possible_pos;
+        score = scores == nullptr ? device_nano<S>() : score;
+        update_score_digest<K, V, S>(bucket_keys_ptr, bucket_capacity, key_pos,
+                                     key, score);
+        atomicAdd(bucket_size_ptr, 1);
+        break;
+      }
+    }
+  }
+  while (occupy_result == OccupyResult::INITIAL) {
+    S* bucket_scores_ptr = BUCKET::scores(bucket_keys_ptr, bucket_capacity, 0);
+    S min_score = MAX_SCORE;
+    int min_pos = -1;
+#pragma unroll
+    for (int j = 0; j < STRIDE_S; j += Load_LEN_S) {
+      __pipeline_memcpy_async(sm_bucket_scores[tx] + j, bucket_scores_ptr + j,
+                              sizeof(S) * Load_LEN_S);
+    }
+    __pipeline_commit();
+    for (int i = 0; i < bucket_capacity; i += STRIDE_S) {
+      if (i < bucket_capacity - STRIDE_S) {
+#pragma unroll
+        for (int j = 0; j < STRIDE_S; j += Load_LEN_S) {
+          __pipeline_memcpy_async(
+              sm_bucket_scores[tx] + diff_buf(i / STRIDE_S) * STRIDE_S + j,
+              bucket_scores_ptr + i + STRIDE_S + j, sizeof(S) * Load_LEN_S);
+        }
+      }
+      __pipeline_commit();
+      __pipeline_wait_prior(1);
+      S temp_scores[Load_LEN_S];
+      S* src = sm_bucket_scores[tx] + same_buf(i / STRIDE_S) * STRIDE_S;
+#pragma unroll
+      for (int k = 0; k < STRIDE_S; k += Load_LEN_S) {
+        *reinterpret_cast<byte16*>(temp_scores) =
+            *reinterpret_cast<byte16*>(src + k);
+#pragma unroll
+        for (int j = 0; j < Load_LEN_S; j += 1) {
+          S temp_score = temp_scores[j];
+          if (temp_score < min_score) {
+            min_score = temp_score;
+            min_pos = i + k + j;
+          }
+        }
+      }
+    }
+    score = scores == nullptr ? device_nano<S>() : score;
```

Comment:
Please check the code of multi-strategy support and refactor them by referring to the[ PR#152](https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/152/) like https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/152/files#diff-271acb18bcae4b5c9972e32f590ac5f2404ee62e7079c1b343c8288161cf8870R106

Reply:
Yes, I support the new evict strategy in the second commit. And I will merge these three commit to one to facilite review.